modules/benchmark/src/main/scala/io/circe/benchmark/FoldingBenchmark.scala

```diff
@@ -0,0 +1,69 @@
+package io.circe.benchmark
+
+import io.circe.{ Json, JsonNumber, JsonObject }
+import java.util.concurrent.TimeUnit
+import org.openjdk.jmh.annotations._
+
+/**
+ * Compare the performance of various ways of folding JSON values.
+ *
+ * The following command will run the benchmarks with reasonable settings:
+ *
+ * > sbt "benchmark/jmh:run -i 10 -wi 10 -f 2 -t 1 io.circe.benchmark.FoldingBenchmark"
+ */
+@State(Scope.Thread)
+@BenchmarkMode(Array(Mode.Throughput))
+@OutputTimeUnit(TimeUnit.SECONDS)
+class FoldingBenchmark extends ExampleData {
+  val doc: Json = Json.arr(intsJson, booleansJson, foosJson, helloWorldJson)
+
+  @Benchmark
+  def withFoldWith: Int = doc.foldWith(
+    new Json.Folder[Int] {
```

Comment:
One cool trick that I learned from Flavio Brasi is embedding this `val`ed functions into their parent scope/class. Somewhat - merging interfaces.

```scala
new Json.Folder[Int] with ((Int, Json) => Int) {
   def apply(i: Int, j: Json): Int = i + j.foldWith(this)
   ...
}
```

This will allocate less.